tools/autograd/gen_variable_type.py

```diff
@@ -169,6 +529,57 @@ def gen_variable_type(declarations, out):
                       if option['has_full_argument_list']]
 
     derivatives = load_derivatives(derivatives_path)
+    deprecated = load_derivatives(deprecated_path)
+
+    def by_name(option):
+        return option['name']
+    def by_aten_name(option):
+        return option.get('aten_name', option['name'])
+
+    aten_decls = sorted(aten_decls, key=by_name)
+    derivatives = sorted(derivatives, key=by_aten_name)
+
+    derivatives_by_signature = {d['signature']: d for d in derivatives}
+    options_by_name = OrderedDict([(k, list(g)) for k, g in groupby(aten_decls, by_name)])
+    options_by_signature = OrderedDict()
+    python_functions = OrderedDict()
+
+    def get_option(derivative):
+        name = derivative.get('aten', derivative['name'])
+        options = options_by_name.get(name, [])
+        if len(options) == 0:
+            raise RuntimeError('Declaration not found for: {}'.format(name))
+        elif len(options) == 1:
+            return options[0]
+        else:
+            raise RuntimeError('ambiguous decl for: {}'.format(name))
+            return None
+
+    for option in aten_decls:
+        args = []
+        for arg in option['arguments']:
+            simple_type = arg['type'].replace(' &', '').replace('const ', '')
+            args.append(simple_type)
+            arg['simple_type'] = simple_type
+        name = option['name']
+        base_name = name[:-1] if option['inplace'] else name
+        signature = '{}({})'.format(base_name, ', '.join(args))
+        if signature not in options_by_signature:
+            options_by_signature[signature] = []
+        options_by_signature[signature].append(option)
```

Comment:
So we need to have a list of options by signature, because of in-place and out-of-place versions? Do they share the same derivative?